[WIP] 6083-mobile-breadcrumb

1Copenut opened the pull request:
## Purpose
To provide contextual breadcrumbs based on `viewport` dimensions on page load, and resize. 

## Methodology
I created two breadcrumb helper JS files, one for static content, one for React components. Each makes use of similar functions to find our breadcrumb markup by ID, clone the list, and determine which one to show based on viewport width. The React version also makes use of lifecycle methods to properly attach and remove breadcrumb methods when component mounts or unmounts.

## Testing
I've tested the breadcrumbs for handling in Chrome, Safari, and Firefox for OSX, and IE11 and Firefox in Windows 7. This included testing with VoiceOver and NVDA screen readers to ensure the breadcrumbs were announcing properly.

## Inclusion in 526 Testing
@goldenmeanie is interested in including these breadcrumbs in the 526 testing that will be happening in the next 2-3 weeks. Please advise if I need to adjust or change things so they're able to be folded into the staging environment.

Commit:
First commit. Have implemented a basic one-step back breadcrumb that triggers on mobile size.

Commit:
[6083] Adding a verified debounced link toggle.

Commit:
Master merge to get up to current.

Commit:
WIP. Adding a workable Metalsmith plugin for transforming breadcrumbs.

Commit:
WIP. Adding further refinement for current page, stronger visiblity rules, and ability to serve via IP address.

Commit:
WIP. Refactoring React apps to handle breadcrumbs in a more universal fashion.

Commit:
WIP. Committing the work on breadcrumbs before pulling master for an audit.

Commit:
Master merge.

Commit:
Commenting out an unneeded const.

Commit:
Adding MD file updates for embedded React apps without their own breadcrumb components.

Commit:
Cleaning up the breadcrumb transformer formatting.

Commit:
Adding all final adjustments to breadcrumb before design review.

Commit:
Editing all the last MD files and React breadcrumb components.

Commit:
Master merge commit

Commit:
Adding final adjustments to unit tests and React breadcrumb helper script.

Commit:
Removing src/js/common/index.js file per Slack conversation.

cvalarida commented on content/includes/breadcrumbs.html:
```diff
@@ -1,39 +1,33 @@
+    {% comment %} Each breadcrumb item has a child object. One of the properties of that object is `file` which loads the front
+    matter metadata from the corresponding Markdown input file. Below, we’ve assigned that object to a variable named 'crumb_child_meta'.
+    If these templates are still using tinyliquid, you can dump the properties of this object using {{ crumb | keys }}. {%
+    endcomment %} {% assign crumb_child_meta = crumb.children.0.file %}
```
The formatting of this block seems kinda wonky. Can we pull that `{% assign... %}` onto its own line and bring `endcomment %}` to the same line as its `{%`?

cvalarida commented on content/includes/breadcrumbs.html:
```diff
@@ -1,39 +1,33 @@
+      </a>
+      {% else %}
+      <a aria-current="page" href="/{{ crumb.path }}">
+        {% if crumb_child_meta.display_title %} {{ crumb_child_meta.display_title }} {% elsif crumb_child_meta.title %} {{ crumb_child_meta.title
```
Readability suffers here by pulling these onto the same line. Can we keep it split up like it was before?

cvalarida reviewed (commented):
I didn't make it very far before I had to bounce to something else, so this isn't a complete review.

What I see so far is looking pretty good. I haven't fired it up myself, but what I remember of it from our call the other day, I like where it's heading.

jbalboni commented on assets/js/breadcrumbs/mobile-breadcrumb.js:
```diff
@@ -0,0 +1,139 @@
+  
+  // Cloned list items and array conversion
+  var targetList = clonedTarget.children;
+  var listArr = Array.prototype.slice.call(targetList);
```
This file isn't going to be processed by Babel, so I think we'll have to stay away from ES6 (or transform this file in the build process).

1Copenut replied:
Yep, I had to go back and change out an ES6 string interpolation this morning. Will commit it in a bit.

jbalboni commented on script/metalsmith/breadcrumbTransformer.js:
```diff
@@ -0,0 +1,22 @@
+
+    $('#va-breadcrumbs').each(() => {
+      $.root().append(
+        '<script type="text/javascript" src="/js/breadcrumbs/mobile-breadcrumb.js"></script>'
```
I wonder if it might be better to inline this script. It's a little long for that, but we are starting to stack up small external JS scripts, which is pushing back the time to first byte on our main bundles.

Commit:
Adding IE11 fixes to breadcrumb helper scripts.

1Copenut commented on assets/js/breadcrumbs/mobile-breadcrumb.js:
```diff
@@ -0,0 +1,139 @@
+var config = {
```
I ran the linter manually a couple of times to be sure, and it didn't throw any errors. Will be sure to hook up the pre-commit hooks so all future runs lint without manual intervention.

1Copenut replied:
Per @jbalboni request, I inlined this script using UglifyJS. I'll remove it here if the inlined version passes muster.

jbalboni replied:
Could we have the breadcrumb transformer run uglify on this script and inline the result? Having the uglified version as the only source is going to be hard to maintain.

jbalboni replied:
Also, I believe the linter ignores files in this directory, since these files aren't run through babel and can't use ES6.

We may want to change this process.

1Copenut commented on content/includes/breadcrumbs.html:
```diff
@@ -1,39 +1,33 @@
-            the corresponding Markdown input file.
+<nav aria-label="Breadcrumb" aria-live="polite" aria-relevant="additions text" class="va-nav-breadcrumbs js-visual {{gatePageClassName}}"
+  id="va-breadcrumbs">
+  <p class="usa-sr-only">Breadcrumb navigation will usually show all page links. It will adjust to show only the previous page when zoomed in, or
```
It borrows from a pattern I saw on an accessible modal dialog. I tried out the screen reader while resizing the viewport, and it was reading out the updated mobile-first breadcrumb, but offering no context. This was my attempt to orient screen reader users on first load. It is easily skipped by tabbing, and can be removed easily if the 508 office says it's not effective.

Commit:
Inlining the breadcrumb script for static content, reformatting breadcrumbs include for readability.

Commit:
Merge branch 'master' into 6083-mobile-breadcrumb

Commit:
Adding UglifyJS to inline static breadcrumb helper, removing an extra ID from discharge wizard link.

Commit:
Updating e2e test for heading change.

Commit:
Master merge

Commit:
Merge branch 'master' into 6083-mobile-breadcrumb

Commit:
Adding a fix for the broken RX e2e test. Need to confirm.

Commit:
Refactored the active RX breadcrumb to show all routes correctly.

jbalboni commented:
@1Copenut It looks like you need to update this to pull in the newer jean-pants package: `yarn upgrade @department-of-veterans-affairs/jean-pants --latest`

jbalboni commented on content/pages/burials-and-memorials/application/530.md:
```diff
@@ -4,13 +4,15 @@ entryname: burials
       <li><a href="/burials-and-memorials/survivor-and-dependent-benefits/"> Survivors and Dependents </a></li>
-      <li><a href="/burials-and-memorials/survivor-and-dependent-benefits/burial-costs/">Handling Burial Costs (Allowances)</a></li>
-    </ul>
+      <li><a aria-current="page" href="/burials-and-memorials/survivor-and-dependent-benefits/burial-costs/">Handling Burial Costs (Allowances)</a></li>
```
This last link isn't the page for the form, it's a content page that's one step back in the breadcrumbs.

jbalboni commented on content/pages/burials-and-memorials/pre-need/form-10007-apply-for-eligibility.md:
```diff
@@ -4,12 +4,14 @@ entryname: pre-need
       <li><a href="/burials-and-memorials/">Burials and Memorials</a></li>
-      <li><a href="/burials-and-memorials/pre-need/">Pre-need Eligibility Determination</a></li>
-    </ul>
+      <li><a aria-current="page" href="/burials-and-memorials/pre-need/">Pre-need Eligibility Determination</a></li>
```
Same as above, this isn't the path to the form page, it's one step back.

1Copenut replied:
@jbalboni There doesn't seem to be a form tied to this one. The content and link goes to a PDF that has to be downloaded or filled out. I'll change this as needed.

jbalboni replied:
Ah, this one's a little confusing since the form isn't launched. We can leave it alone and update it whenever it goes live (a couple months, probably).

1Copenut replied:
👍

jbalboni commented on content/pages/education/apply-for-education-benefits/application/1990.md:
```diff
@@ -5,12 +5,15 @@ layout: page-react.html
-      <li class="parent"><a href="/education/apply-for-education-benefits/">Apply for Education Benefits</a></li>
-    </ul>
+      <li><a href="/education/apply/">Application Process</a></li>
+      <li><a aria-current="page"  href="/education/apply-for-education-benefits/application/1990/introduction">Apply for Education Benefits</a></li>
```
As you go through the form, this link is going to be inaccurate, since the introduction page is just the first page of the form. I think our typical pattern in the site is for breadcrumbs to not include a link to the current page. Are we changing that? If we are, we may need to move the breadcrumbs into React component for all the React apps, so we can keep them up to date.

1Copenut replied:
@jbalboni Good call. The final breadcrumb link is being put back, but we're using CSS to remove `pointer-events`, and no, the breadcrumb should not change as you move through multi-step forms. I'll update this href to `/education/apply-for-education-benefits/application/1990`

jbalboni commented on script/metalsmith/breadcrumbTransformer.js:
```diff
@@ -0,0 +1,26 @@
+
+    const data = files[file];
+    const $ = cheerio.load(data.contents.toString());
+    const minifiedCode = UglifyJS.minify(breadcrumb);
```
Might be useful to minify this outside of the `forEach` loop, since the same minified code is used for each file.

jbalboni commented on src/js/claims-status/components/Breadcrumbs.jsx:
```diff
@@ -1,14 +1,59 @@
+    });
+  }
+
+  componentDidUpdate() {
```
This method may run multiple times, should we be adding new event listeners each time? It seems like adding them in `componentDidMount` would be enough.

jbalboni commented on src/js/claims-status/components/Breadcrumbs.jsx:
```diff
@@ -1,14 +1,59 @@
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('DOMContentLoaded', () => {
```
I don't think this is removing an event listener, since you're passing in a new function that won't match any of the current event listeners. I think what we usually to is create a method on a component and then do `addEventListener('resize', this.toggleLinks)` and `removeEventListener('resize', this.toggleLinks)`

ncksllvn replied:
☝️ ➕

jbalboni commented on src/js/health-records/components/Breadcrumbs.jsx:
```diff
@@ -12,14 +36,28 @@ class Breadcrumbs extends React.Component {
     if (pathname.match(/download\/?$/)) {
       crumbs.push(<Link to="/" key="main">Get Your VA Health Records</Link>);
-      crumbs.push(<span key="download"><strong>Download Your Health Records</strong></span>);
+      crumbs.push(<Link to="/" key="download">Download Your VA Health Records</Link>);
```
Is this an intentional heading change?

jbalboni commented on src/js/utils/breadcrumb-helper.js:
```diff
@@ -0,0 +1,169 @@
+const config = {
```
This file should move to `src/platform/utilities/ui/`

jbalboni commented on src/js/utils/breadcrumb-helper.js:
```diff
@@ -0,0 +1,169 @@
+const config = {
```
I'm a little concerned that we're doing this much direct DOM manipulation in a React component. I realize you're reusing code that's also being used for the static pages that isn't in React, but given that we're already going to have maintain two slightly different versions of this code, I wonder if it might be better to have a React breadcrumbs component that is written in idiomatic React and use that in the React apps that need it. I'm pretty sure the React code for this would much simpler than the DOM version here.

ncksllvn replied:
☝️ ➕

jbalboni commented on src/sass/base/_b-utils.scss:
```diff
@@ -0,0 +1,86 @@
+.full-width {
```
I think you can delete this file.

jbalboni commented on src/sass/base/_va.scss:
```diff
@@ -0,0 +1,1497 @@
+html,
```
I think you can delete this file, too.

ncksllvn commented on src/js/claims-status/containers/AdditionalEvidencePage.jsx:
```diff
@@ -64,23 +64,33 @@ class AdditionalEvidencePage extends React.Component {
+
+    const listItem = breadcrumbArr.map(item => {
+      return <li key={item.key}><a href={item.href}>{item.text}</a></li>;
+    });
```
This is kinda picky, but `listItem` seems vague to me. Maybe just `breadcrumbItems` or something?

ncksllvn commented on src/js/discharge-wizard/components/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,48 @@
+    window.removeEventListener('resize', () => {
+      debouncedToggleLinks.bind(this);
+    });
+  }
```
Does each application have its own breadcrumb component? Are there differences between them that can't be consolidated via props?

ncksllvn reviewed (commented):
Left a few comments, but I'm a concerned that there is too much code redundancy and I think we could improve by consolidating breadcrumbs logic into a common React component. We could even pass the list items into the React root from static pages.

What are your all's thoughts on that? It just seems like solid use case for Jean Pants to me 👖

Commit:
Refactoring a single Breadcrumb component for eventual Design System inclusion

Commit:
Master merge.

Commit:
Refactoring breadcrumb to use single, simpler component.

Commit:
Merge branch 'master' into 6083-mobile-breadcrumb

Commit:
Moving the prescriptions actions back to master, and simpler logic for prescriptions breadcrumbs.

Commit:
Merge branch 'master' into 6083-mobile-breadcrumb

Commit:
Adding Cheerio, Uglify to dev dependencies.

Commit:
Pulling latest design system before testing.

Commit:
Removing old breadcrumbs in favor of single component. Updating link URLs for correctness in MD files.

1Copenut commented:
I changed this to WIP because we're testing this week or early next, and didn't know what the right environment was. I'll follow up with teams to ensure we get it in the right place(s).

The `<Breadcrumb/>` component has been refactored out to a single, and made more robust. I'll take it as a follow on item to move this component to the Design System after user research is done.

jbalboni commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,118 @@
+      mobileShow: false,
+    };
+
+    this.id = this.props.id || uniqid('va-breadcrumbs-');
```
This is a little risky in that props could change and the instance variables wouldn't change to reflect that. You could make these `const` variables inside the render method (or other methods where they're needed), I think.

jbalboni commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,118 @@
+    this.mobileWidth = this.props.mobileWidth || 425;
+
+    this.debouncedToggleDisplay = this.debouncedToggleDisplay.bind(this);
+    this.renderBreadcrumbLinks = this.renderBreadcrumbLinks.bind(this);
```
We do this both ways in our codebase, but I think the current preference is to use the auto bind syntax:

```js
someMethod = () => {
  ...
}
```

jbalboni commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,118 @@
+        );
+      }
+
+      return <li>{React.cloneElement(child)}</li>;
```
Is the cloneElement call necessary here?

Commit:
Adding final fixes to URLs for breadcrumbs.

jbalboni commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,118 @@
+}
+
+Breadcrumbs.propTypes = {
+  crumbs: PropTypes.array,
```
Doesn't look like we use this?

1Copenut replied:
No, I'll remove this momentarily. Legacy item.

jbalboni commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,118 @@
+
+  render() {
+    const {
+      ...breadcrumbProps
```
I'm not sure I understand why this is being done.

1Copenut replied:
@jbalboni There's a few instances where custom props are being fed into the `<Breadcrumb>` component, like the Facility Locator. I added this `...breadcrumbProps` defensively to ensure those could be added as necessary. Will remove if it's not something we want here.

jbalboni replied:
Ok. It looks like you could `...this.props` later in this method to get the same effect without this line, unless I'm missing something.

1Copenut replied:
@jbalboni Some instances of the `<Breadcrumb>` have custom props passed down, like the Facility Locator. I'm not sure they're needed, but didn't want to remove them either. Added this as a defensive way to not break things. Can remove as needed.

jbalboni commented:
@1Copenut What's the plan for how this would be tested, do you know? If it's merged it'll go to production, which is maybe not ideal, unless we can make the changes behind a flag somehow (that seems difficult with all the changes here). We can publish arbitrary branches to dev.vets.gov, so that might be your best option, but I'm not sure if we can keep them from getting overwritten.

Commit:
Changing to Lodash.uniqueid and refactoring Breadcrumb component for current practices.

Commit:
Fixing merge conflict before starting PR review.

Commit:
Adding two URL fixes for breadcrumbs

ncksllvn commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,116 @@
+  listId: PropTypes.string,
+  mobileWidth: PropTypes.number,
+};
+
```
Great work!! I wonder if there's any way we could use this on pages that also aren't registered as React applications, like we do with the navigation. It's all solid code but that way we wouldn't have to worry about maintaining a separate JS file that performs the same function minus React. @jbalboni do you think it would be okay to do something like that in our common startup code?

ncksllvn commented on assets/js/breadcrumbs/mobile-breadcrumb.js:
```diff
@@ -0,0 +1,129 @@
+  mobileClass: 'va-nav-breadcrumbs-list__mobile-link',
+  triggerDelay: 500,
+  triggerWidth: 425,
+};
```
I wonder if we could have `triggerWidth` be derived from what CSS has set as the small device width. Probably just wishful thinking.

1Copenut replied:
@ncksllvn I like this suggestion a lot. I've been looking through the CSS and found 481px was the small-screen SCSS variable. I'm going to make that my breakpoint width for these breadcrumbs for consistency.

ncksllvn commented on assets/js/breadcrumbs/mobile-breadcrumb.js:
```diff
@@ -0,0 +1,139 @@
+  var breadcrumb = document.getElementById(targetId);
+  var clone = document.getElementById(targetId + '-clone');
+
+  if (window.innerWidth <= config.triggerWidth) {
```
`platform/utilities/accessibility` detects screen size using `matchMedia`. Might be worth doing something with here too.

ncksllvn commented on src/platform/utilities/ui/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,116 @@
+
+  componentDidMount() {
+    const mobileWidth = this.props.mobileWidth || 425;
+
```
`425` would make sense as a `defaultProp`

ncksllvn reviewed (commented):
Solid improvements, well done! I'm curious about what the rollout process for this will be as well. Since it has such a wide impact I think it makes sense to have some standalone unit and E2E tests written for the Breadcrumb component. I think it would also be interesting to try to explore ideas around rendering breadcrumbs always as React component, even for those that don't have a dedicated React entry point, just so we don't have maintain two separate files of complex logic.

Commit:
Adding a full breadcrumb unit test.

Commit:
Adding a first e2e breadcrumb test for GI Bill Comparison Tool

Commit:
Adding breadcrumb end to end tests that reflect depth and CSS properties.

Commit:
Master merge.

Commit:
Updating directory structure to support all tests.

Commit:
Updating to unordered lists, revising tests, adding rollout plan revisions.

1Copenut commented:
@jbalboni @ncksllvn 
I added in the feedback you proposed to solidify our breakpoint around the CSS media query, and to use `window.mediaMatch`. Wasn't familiar with that, so it was a good thing to line up with other code in our base.

Added a full unit test and 10 or 11 e2e tests specifically for the breadcrumbs on different React apps. 

Added a rollout plan MD file here: https://github.com/department-of-veterans-affairs/vets-website/pull/7579/files#diff-22661d7a4cd7454a743b7a592695bf88

Commit:
Adding a missed file update for education benefit MD file.

1Copenut commented:
Closing because this work is superseded by https://github.com/department-of-veterans-affairs/vets-website/pull/8004
